Add EventBridge event source

pmuens opened the pull request:
## What did you implement:

Closes #6363

Adds support for the new AWS EventBridge service (see discussion in issue mentioned above for more 
details).

## How did you implement it:

Added the logic via raw SDK calls which are wrapped in a custom CloudFormation resource.

## How can we verify it:

Use the following `serverless.yml` file:

```yml
service: test

provider:
  name: aws
  runtime: nodejs10.x
  versionFunctions: false

functions:
  test:
    handler: handler.hello
    events:
      - eventBridge:
          schedule: rate(1 minute)
      - eventBridge:
          # creating a new event bus
          eventBus: custom-event-bus
          pattern:
            source:
              - serverless.test
      - eventBridge:
          # re-using an existing event bus
          eventBus: arn:aws:events:us-east-1:12345:event-bus/existing-event-bus
          pattern:
            source:
              - serverless.test
```

**NOTE:** You can find more config options in the service definition of https://github.com/serverless/serverless/issues/6363#issuecomment-510853941

## Todos:

_**Note: Run `npm run test-ci` to run all validation checks on proposed changes**_

- [x] Write tests and confirm existing functionality is not broken.  
       **Validate via `npm test`**
- [x] Write documentation
- [x] Ensure there are no lint errors.  
       **Validate via `npm run lint-updated`**  
       _Note: Some reported issues can be automatically fixed by running `npm run lint:fix`_
- [x] Ensure introduced changes match Prettier formatting.  
       **Validate via `npm run prettier-check-updated`**  
       _Note: All reported issues can be automatically fixed by running `npm run prettify-updated`_
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

**_Is this ready for review?:_** YES
**_Is it a breaking change?:_** NO

dschep commented on lib/plugins/aws/customResources/index.js:
```diff
@@ -1,9 +1,21 @@
 const path = require('path');
+const BbPromise = require('bluebird');
 const _ = require('lodash');
+const fse = BbPromise.promisifyAll(require('fs-extra'));
```
`fs-extra` has native promise support.

```suggestion
const fse = require('fs-extra');
```

pmuens replied:
Good catch. Unfortunately we're using a super old version of `fs-extra` and updating looked quite scary so I used the promisified solution.

dschep commented on lib/plugins/aws/customResources/index.js:
```diff
@@ -1,9 +1,21 @@
 const createZipFile = require('../../../utils/fs/createZipFile');
 
+function copyCustomResources(srcDirPath, destDirPath) {
+  return fse.ensureDirAsync(destDirPath).then(() => fse.copyAsync(srcDirPath, destDirPath));
```
```suggestion
  return fse.ensureDir(destDirPath).then(() => fse.copy(srcDirPath, destDirPath));
```

dschep commented on lib/plugins/aws/customResources/resources/eventBridge/handler.js:
```diff
@@ -0,0 +1,125 @@
+    eventBus: EventBridgeConfig.EventBus,
+    ruleName: EventBridgeConfig.RuleName,
+  })
+    .then(() =>
```
I know we can't use `async`/`await` in sls code itself, but can't a lambda use it since we control the runtime it's deploying with?

pmuens replied:
Yes, I did the same in the intial PR for existing S3 bucket. @medikoo pointed out that there's some potential rejected promises might not be propagated correctly and he advised that I should use plain Promises rather than `async` / `await`.

medikoo replied:
My original point was that we should use _callback_ resolution instead of _thenable_ lambda resolution, as in latter case AWS env doesn't wait for orphaned async flows to end, and they may leak to further invocations.

Technically we can use async/await as long as top level lambda handler function is not an async function, and we propagate async functions result to AWS callback.

dschep commented on lib/plugins/aws/lib/naming.js:
```diff
@@ -485,4 +485,16 @@ module.exports = {
   },
+  // Event Bridge
+  getCustomResourceEventBridgeHandlerFunctionName() {
+    return 'custom-resource-event-bridge';
```
does this not cause issues if you try to use the custom resource in multiple projects?

pmuens replied:
🤔 it shouldn't since the handler is only defined once per service.

dschep replied:
oh. handler function name. i was thinking lambda name.

dschep commented on lib/plugins/aws/package/compile/events/eventBridge/index.js:
```diff
@@ -0,0 +1,155 @@
+      if (functionObj.events) {
+        functionObj.events.forEach((event, idx) => {
+          if (typeof event.eventBridge === 'object') {
+            idx++;
```
Is this just because we don't want `FuncName-rule-0`?

pmuens replied:
Yes, exactly.

dschep commented on lib/plugins/aws/package/compile/events/eventBridge/index.js:
```diff
@@ -0,0 +1,155 @@
+
+      iamRoleStatements.push({
+        Effect: 'Allow',
+        Resource: `arn:aws:lambda:*:*:function:${FunctionName}`,
```
shouldn't we set the region & account id?

pmuens replied:
Oh yeah, I tried to do that but getting access to the account id is quite tedious. I can look into that tomorrow.

dschep replied:
```suggestion
        Resource: {
          'Fn::Join': [
            ':',
            [
              'arn:aws:lambda',
              { Ref: 'AWS::Region' },
              { Ref: 'AWS::AccountId' },
              'function',
              FunctionName,
            ]
          }
```

pmuens replied:
![image](https://user-images.githubusercontent.com/1606004/61823843-e1cc5600-ae5c-11e9-97ab-20981ba9ffb1.png)

Thanks! 👌

That happens when you spend too much time in the programmatic realm.

dschep replied:
I didn't remember it, I just knew I had projects that did similar and where to look: https://github.com/dschep/bikehero/blob/master/serverless.yml#L17-L22

pmuens replied:
> knowing where to find an answer can be just as valuable as knowing the answer 🙂

👌 💯

pmuens commented:
@dschep thanks for your review! I just updated the PR :+1:

Let me know if there's anything else I need to do here.

pmuens commented:
@dschep can you approve this PR again? I rebased it and found a minor bug where the `eventBridge` event could not be used together with another event source. I added a regression test for that case and tested everything thoroughly.

Thanks in advance!

pmuens commented:
Thanks @dschep :+1:
